CMakeLists: Avoid type errors in probes

The timespec_get function expects a struct timespec * argument,
not a struct tm * argument.

Do not attempt to return a char * from main, which returns type int.

Submitted upstream: <https://github.com/fluent/fluent-bit/pull/6704>

diff --git a/CMakeLists.txt b/CMakeLists.txt
index c6315bb2a..ed3c091ae 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -738,7 +738,7 @@ endif()
 check_c_source_compiles("
   #include <time.h>
   int main() {
-     struct tm tm;
+     struct timespec tm;
      return timespec_get(&tm, TIME_UTC);
   }" FLB_HAVE_TIMESPEC_GET)
 if(FLB_HAVE_TIMESPEC_GET)
diff --git a/lib/cmetrics/CMakeLists.txt b/lib/cmetrics/CMakeLists.txt
index 89089c0f2..dcee1ea1b 100644
--- a/lib/cmetrics/CMakeLists.txt
+++ b/lib/cmetrics/CMakeLists.txt
@@ -94,7 +94,7 @@ include_directories(
 check_c_source_compiles("
   #include <time.h>
   int main() {
-     struct tm tm;
+     struct timespec tm;
      return timespec_get(&tm, TIME_UTC);
   }" CMT_HAVE_TIMESPEC_GET)
 if(CMT_HAVE_TIMESPEC_GET)
@@ -107,7 +107,7 @@ check_c_source_compiles("
   int main() {
      struct tm tm;
      struct timespec tms;
-     return gmtime_r(&tms.tv_sec, &tm);
+     return gmtime_r(&tms.tv_sec, &tm) != 0;
   }" CMT_HAVE_GMTIME_R)
 if(CMT_HAVE_GMTIME_R)
   CMT_DEFINITION(CMT_HAVE_GMTIME_R)
